[Storage] Continue storage deletion in `sky storage delete` when some fail

andylizf opened the pull request:
<!-- Describe the changes in this PR -->

Fix #4050.

Make `sky storage delete` continue deleting remaining storages when some fail, instead of aborting the entire operation. Failed deletions are reported individually.

We Implemented this feature via adding a `continue_on_error` parameter in `run_in_parallel`

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

andylizf commented:
@cblmemo PTAL, thanks!

andylizf commented:
I inserted five new records into the 'global_user_state' database: three records containing the previous storage handle value, and two records with null handle values.

```bash
andyl@DESKTOP-7FP6SMO ~/skypilot (fix-storage-recover) [1]> sky storage dele
te -a
Deleting 6 storages: skypilot-workdir-andyl-74d418df, test-storage-1, test-storage-2, test-storage-3, test-storage-4, test-storage-5. Proceed? [Y/n]: 
Error deleting storage test-storage-3: In global_user_state, storage name 'test-storage-3' does not match handle.storage_name 'skypilot-workdir-andyl-577894a6'
Error deleting storage test-storage-2: In global_user_state, storage name 'test-storage-2' does not match handle.storage_name 'skypilot-workdir-andyl-577894a6'
Error deleting storage test-storage-4: Storage name 'test-storage-4' not found.
Error deleting storage test-storage-1: In global_user_state, storage name 'test-storage-1' does not match handle.storage_name 'skypilot-workdir-andyl-577894a6'
Error deleting storage test-storage-5: Storage name 'test-storage-5' not found.
Deleted GCS bucket skypilot-workdir-andyl-74d418df.```

cblmemo commented on sky/cli.py:
```diff
@@ -3543,7 +3542,15 @@ def storage_delete(names: List[str], all: bool, yes: bool):  # pylint: disable=r
                 show_default=True)
 
-    subprocess_utils.run_in_parallel(sky.storage_delete, names)
+    def delete_storage(name: str):
```
```suggestion
    def delete_storage(name: str) -> None:
```

cblmemo commented on sky/cli.py:
```diff
@@ -3543,7 +3542,15 @@ def storage_delete(names: List[str], all: bool, yes: bool):  # pylint: disable=r
+    def delete_storage(name: str):
+        try:
+            sky.storage_delete(name)
+        except Exception as e:  # pylint: disable=broad-except
```
can we only except certain types of exceptions?

cblmemo commented on sky/global_user_state.py:
```diff
@@ -826,8 +826,13 @@ def get_storage_names_start_with(starts_with: str) -> List[str]:
     return [row[0] for row in rows]
 
 
+def get_storage_names() -> List[str]:
+    rows = _DB.cursor.execute('SELECT name FROM storage')
+    return [row[0] for row in rows]
```
can we use sth like `get_glob_storage_name('*')` to replace this function?

andylizf replied:
Good catch!

cblmemo commented on sky/utils/subprocess_utils.py:
```diff
@@ -96,29 +96,47 @@ def get_parallel_threads(cloud_str: Optional[str] = None) -> int:
+        arguments. If continue_on_error=True, failed operations will have
+        their exceptions in the result list.
     """
-    # Reference: https://stackoverflow.com/questions/25790279/python-multiprocessing-early-termination # pylint: disable=line-too-long
```
why not removing this reference

andylizf replied:
You mean why remove the reference? I think it's trivial now, and the reference doesn't add much value to the code.

cblmemo commented on sky/utils/subprocess_utils.py:
```diff
@@ -96,29 +96,47 @@ def get_parallel_threads(cloud_str: Optional[str] = None) -> int:
                     args: Iterable[Any],
-                    num_threads: Optional[int] = None) -> List[Any]:
+                    num_threads: Optional[int] = None,
+                    continue_on_error: bool = False) -> List[Any]:
```
adding this seems change the return value type from `Value` to `Union[Value, Exception]`, which looks a little bit strange to me. Why do we need this anyway? aren't we already add the try-except in the func?

andylizf replied:
You are right. I thought about not adding a function to catch and log these exceptions, but instead handle all exceptions in `run_in_parallel`. Anyway, it made sense to have this function have an argument like `continue_on_error`.
But then I realized it violated the abstraction concept and would make it harder to print custom logs. So I ended up wrapping the function and logging it.

cblmemo reviewed (commented):
Thanks for the prompt fix @andylizf ! left some discussions.

cblmemo commented on sky/cli.py:
```diff
@@ -3525,11 +3525,10 @@ def storage_delete(names: List[str], all: bool, yes: bool):  # pylint: disable=r
     if all:
-        storages = sky.storage_ls()
-        if not storages:
+        names = global_user_state.get_glob_storage_name('*')
```
```suggestion
        # Use '*' to get all storages.
        names = global_user_state.get_glob_storage_name(cluster_name='*')
```
nit: use kwargs for literal.

cblmemo commented on sky/cli.py:
```diff
@@ -3543,7 +3542,18 @@ def storage_delete(names: List[str], all: bool, yes: bool):  # pylint: disable=r
+    def delete_storage(name: str) -> None:
+        try:
+            sky.storage_delete(name)
+        except (exceptions.StorageBucketDeleteError, PermissionError) as e:
+            click.secho(f'Error deleting storage {name}: {e}', fg='red')
+        except ValueError as e:
+            click.secho(f'Error deleting storage {name}: {e}', fg='red')
```
```suggestion
        except (exceptions.StorageBucketDeleteError, PermissionError, ValueError) as e:
            click.secho(f'Error deleting storage {name}: {e}', fg='red')
```
can we merge them?

cblmemo replied:
Actually, seems like here is one exception that is not being caught. After a second thought, for simplicity, we can probably just do broad except here..

```bash
(skypilot) (base) ➜  skypilot git:(remove_token) ✗ sky storage delete -a                                                   
Deleting 15 storages: skypilot-workdir-windsey-afa32d67, skypilot-workdir-windsey-81d2115f, skypilot-workdir-windsey-b8fd0382, skypilot-workdir-windsey-499178bb, skypilot-workdir-windsey-dd64cc00, skypilot-workdir-windsey-d04e17a8, skypilot-workdir-windsey-43925bf5, skypilot-workdir-windsey-3375424a, skypilot-workdir-windsey-5e824f17, skypilot-workdir-windsey-db755382, skypilot-workdir-windsey-25270ec6, skypilot-workdir-windsey-1cb4689a, skypilot-workdir-windsey-f40df433, skypilot-workdir-windsey-2ef8a0bf, skypilot-workdir-windsey-58878e24. Proceed? [Y/n]: 
No backing stores found. Deleting storage.
botocore.exceptions.ClientError: An error occurred (403) when calling the HeadBucket operation: Forbidden

The above exception was the direct cause of the following exception:

sky.exceptions.StorageBucketGetError: Failed to access existing bucket 'skypilot-workdir-windsey-afa32d67'. This is likely because it is a private bucket you do not have access to.
To fix: 
  1. If you are trying to create a new bucket: use a different name.
  2. If you are trying to connect to an existing bucket: make sure your cloud credentials have access to it. To debug, consider running `aws s3 ls skypilot-workdir-windsey-afa32d67`.
```

andylizf replied:
Yeah... I agree

cblmemo commented on sky/cli.py:
```diff
@@ -3543,7 +3542,18 @@ def storage_delete(names: List[str], all: bool, yes: bool):  # pylint: disable=r
+            click.secho(f'Error deleting storage {name}: {e}', fg='red')
+        except ValueError as e:
+            click.secho(f'Error deleting storage {name}: {e}', fg='red')
+        except Exception as e:  # pylint: disable=broad-except
+            with ux_utils.print_exception_no_traceback():
+                raise e
+
```
```suggestion

```
It seems like we dont need it? We can simply not `except` it. And any exceptions raised by `sky.storage_delete` should already have the surpress traceback context?

andylizf replied:
Why they should have the traceback context supressed?

cblmemo replied:
Cuz this is a user-facing API and we want to keep the error message simple. The content in the error message should be self-explanatory.

cblmemo reviewed (commented):
Thanks @andylizf ! Mostly looks good to me. After those nits it should be ready to go!

cblmemo reviewed (approved):
Thanks @andylizf ! LGTM. Please lmk when you pass the test and it should be ready to go!

cblmemo commented:
@andylizf are you pushing to the wrong branch

cblmemo commented:
seems like there are a conflict with the master branch. Could you help fix it?

Also, how is the testing goes? If it passed then we can merge it.

andylizf commented:
Now it works.

```
andyl@Andys-MacBook-Pro ~/P/skypilot (fix-storage-recover)> sky storage delete -a                                                          (skypilot) 
Deleting 10 storages: skypilot-workdir-andyl-01fca537, skypilot-workdir-andyl-1904291f, skypilot-workdir-andyl-5707be21, skypilot-workdir-andyl-b078e3e3, skypilot-workdir-andyl-b5298d10, test-storage-1, test-storage-2, test-storage-3, test-storage-4, test-storage-5. Proceed? [Y/n]: 
Error deleting storage test-storage-5: Storage name 'test-storage-5' not found.
Error deleting storage test-storage-1: In global_user_state, storage name 'test-storage-1' does not match handle.storage_name 'skypilot-workdir-andyl-b078e3e3'
Error deleting storage test-storage-3: In global_user_state, storage name 'test-storage-3' does not match handle.storage_name 'skypilot-workdir-andyl-b078e3e3'
Error deleting storage test-storage-4: Storage name 'test-storage-4' not found.
Error deleting storage test-storage-2: In global_user_state, storage name 'test-storage-2' does not match handle.storage_name 'skypilot-workdir-andyl-b078e3e3'
Deleted GCS bucket skypilot-workdir-andyl-b5298d10.
Deleted GCS bucket skypilot-workdir-andyl-1904291f.
Deleted GCS bucket skypilot-workdir-andyl-5707be21.
Deleted GCS bucket skypilot-workdir-andyl-01fca537.
Deleted GCS bucket skypilot-workdir-andyl-b078e3e3.
```

cblmemo commented:
> Now it works.
> 
> ```
> andyl@Andys-MacBook-Pro ~/P/skypilot (fix-storage-recover)> sky storage delete -a                                                          (skypilot) 
> Deleting 10 storages: skypilot-workdir-andyl-01fca537, skypilot-workdir-andyl-1904291f, skypilot-workdir-andyl-5707be21, skypilot-workdir-andyl-b078e3e3, skypilot-workdir-andyl-b5298d10, test-storage-1, test-storage-2, test-storage-3, test-storage-4, test-storage-5. Proceed? [Y/n]: 
> Error deleting storage test-storage-5: Storage name 'test-storage-5' not found.
> Error deleting storage test-storage-1: In global_user_state, storage name 'test-storage-1' does not match handle.storage_name 'skypilot-workdir-andyl-b078e3e3'
> Error deleting storage test-storage-3: In global_user_state, storage name 'test-storage-3' does not match handle.storage_name 'skypilot-workdir-andyl-b078e3e3'
> Error deleting storage test-storage-4: Storage name 'test-storage-4' not found.
> Error deleting storage test-storage-2: In global_user_state, storage name 'test-storage-2' does not match handle.storage_name 'skypilot-workdir-andyl-b078e3e3'
> Deleted GCS bucket skypilot-workdir-andyl-b5298d10.
> Deleted GCS bucket skypilot-workdir-andyl-1904291f.
> Deleted GCS bucket skypilot-workdir-andyl-5707be21.
> Deleted GCS bucket skypilot-workdir-andyl-01fca537.
> Deleted GCS bucket skypilot-workdir-andyl-b078e3e3.
> ```

Thanks! LGTM. Merging now
